pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentmetadata/SegmentZkMetadataFetcher.java

```diff
@@ -0,0 +1,121 @@
+ */
+public class SegmentZkMetadataFetcher {
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private final String _segmentZKMetadataPathPrefix;
+  private final List<SegmentZkMetadataFetchListener> _registeredListeners;
+  private final Set<String> _onlineSegmentsCached;
+
+  private boolean _initialized;
+
+  public SegmentZkMetadataFetcher(String tableNameWithType, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableNameWithType = tableNameWithType;
+    _propertyStore = propertyStore;
+    _segmentZKMetadataPathPrefix = ZKMetadataProvider.constructPropertyStorePathForResource(tableNameWithType) + "/";
+    _registeredListeners = new ArrayList<>();
+    _onlineSegmentsCached = new HashSet<>();
+    _initialized = false;
+  }
+
+  public void init(IdealState idealState, ExternalView externalView, Set<String> onlineSegments) {
+    if (!_initialized) {
+      _initialized = true;
+      if (!_registeredListeners.isEmpty()) {
+        // Bulk load partition info for all online segments
+        int numSegments = onlineSegments.size();
+        List<String> segments = new ArrayList<>(numSegments);
+        List<String> segmentZKMetadataPaths = new ArrayList<>(numSegments);
+        for (String segment : onlineSegments) {
+          segments.add(segment);
+          segmentZKMetadataPaths.add(_segmentZKMetadataPathPrefix + segment);
+        }
+        _onlineSegmentsCached.addAll(onlineSegments);
+        List<ZNRecord> znRecords = _propertyStore.get(segmentZKMetadataPaths, null, AccessOption.PERSISTENT, false);
+        for (SegmentZkMetadataFetchListener listener : _registeredListeners) {
+          listener.init(idealState, externalView, segments, znRecords);
+        }
+      }
+    } else {
+      throw new RuntimeException("Segment zk metadata fetcher has already been initialized!");
+    }
+  }
+
+  public void registerListener(SegmentZkMetadataFetchListener listener) {
+    if (!_initialized) {
+      _registeredListeners.add(listener);
+    } else {
+      throw new RuntimeException("Segment zk metadata fetcher has already been initialized! "
+          + "Unable to register more listeners.");
+    }
+  }
+
+  public List<SegmentZkMetadataFetchListener> getRegisteredListeners() {
+    return _registeredListeners;
+  }
+
+  public synchronized void onAssignmentChange(IdealState idealState, ExternalView externalView,
+      Set<String> onlineSegments) {
+    if (!_registeredListeners.isEmpty()) {
+      int numSegments = onlineSegments.size();
+      List<String> segments = new ArrayList<>(numSegments);
+      List<String> segmentZKMetadataPaths = new ArrayList<>(numSegments);
+      for (String segment : onlineSegments) {
+        if (!_onlineSegmentsCached.contains(segment)) {
+          segments.add(segment);
+          segmentZKMetadataPaths.add(_segmentZKMetadataPathPrefix + segment);
+        }
+      }
+      _onlineSegmentsCached.addAll(onlineSegments);
```

Comment:
(minor) Can be simplified
```suggestion
      for (String segment : onlineSegments) {
        if (_onlineSegmentsCached.add(segment)) {
          segments.add(segment);
          segmentZKMetadataPaths.add(_segmentZKMetadataPathPrefix + segment);
        }
      }
```